codec/src/codec.rs

```diff
@@ -125,9 +135,6 @@ pub trait Reader {
 
 /// Trait for codec write operations
 pub trait Writer {
-    /// Writes a value of type T
-    fn write<T: Codec>(&mut self, value: &T);
-
```

Comment:
Allowed a client to easily accidentally write infinite recursion code that ends up with a stack overflow:
```
impl Codec for X {
    fn write(&self, writer: &mut impl Writer);
        writer.write(self);
    }
}
```

instead, force use of `Z.write(writer);`. If you tried `self.write(writer)` inside of the `self.write()` function, the linter would catch the recursion issue